ci: Add IWYU job

hebasto opened the pull request:
This PR separates the IWYU checks into its own CI job to provide faster feedback to developers. No other changes are made to the treatment of IWYU warnings. The existing “tidy” CI job will no longer run IWYU.

See also the discussion of https://github.com/bitcoin/bitcoin/pull/33779, specifically this [comment](https://github.com/bitcoin/bitcoin/pull/33779#issuecomment-3491515263):
> Maybe a better approach would be to run the enforced sections in a separate, faster job? Some of the linters are already a bit annoying to invoke locally, so I usually just run the lint job. Doing the same for the includes seems fine to me.

Based on ideas from https://github.com/bitcoin/bitcoin/pull/32953.

w0xlt reviewed (commented):
Concept ACK

maflcko commented on ci/test/03_test_script.sh:
```diff
@@ -44,7 +44,7 @@ echo "=== END env ==="
 # Don't apply patches in the tidy job, because it relies on the `git diff`
```
```suggestion
# Don't apply patches in the iwyu job, because it relies on the `git diff`
```

nit, if you re-touch

maflcko reviewed (approved):
lgtm. Seems fine

maflcko commented:
> instructions

Some basic instructions could be added after the `echo "^^^ ⚠️ Failure generated from IWYU"`, if needed.

maflcko commented on ci/test/03_test_script.sh:
```diff
@@ -213,7 +217,9 @@ if [ "${RUN_TIDY}" = "true" ]; then
+fi
 
+if [ "${RUN_IWYU}" == "true" ]; then
   # TODO: Consider enforcing IWYU across the entire codebase.
```
Maybe:

```sh
echo "^^^ ⚠️ Failure generated from IWYU"
echo
echo "Some adjustments to the diff may be needed:"
echo
echo "* Use the C++ headers: E.g. <cerrno> over <errno.h>"
echo '* Use #include <__.h> over #include "__.h" for all headers'
echo "* Sort the headers, so that they are in the right pre-existing section, if"
echo "  there are any."

maflcko reviewed (approved):
lgtm

fanquake commented on ci/test/03_test_script.sh:
```diff
@@ -234,6 +240,13 @@ if [ "${RUN_TIDY}" = "true" ]; then
+    echo "Some adjustments to the diff may be needed:"
+    echo
+    echo "* Use the C++ headers: E.g. <cerrno> over <errno.h>"
+    echo '* Use #include <__.h> over #include "__.h" for all headers'
```
I'm wondering, give we are building IWYU from source, can we just patch it, so the output is correct (uses <>); avoiding the need for this fixup.

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -225,6 +231,12 @@ if [ "${RUN_TIDY}" = "true" ]; then
+    echo
+    echo "Some adjustments to the diff may be needed:"
+    echo
+    echo "* Use the C++ headers: E.g. <cerrno> over <errno.h>"
```
> nit: Looking at the output, I think this is already done correctly?

I don't think so. The generated diffs still [suggest](https://github.com/bitcoin/bitcoin/actions/runs/19938614368/job/57170256836) the C headers:
```diff
+#include <errno.h>
```

maflcko replied:
I see. I guess they could also be fixed. Either with a patch in iwyu, since it is already patched, or with `modernize-deprecated-headers` clang-tidy-diff, or with a stand-alone mapping script: `./ci/test/modernize-deprecated-headers.py $( git diff --name-only )`

For ref, the mapping is probably:

```
<assert.h>:<cassert>
<ctype.h>:<cctype>
<errno.h>:<cerrno>
<float.h>:<cfloat>
<limits.h>:<climits>
<locale.h>:<clocale>
<math.h>:<cmath>
<setjmp.h>:<csetjmp>
<signal.h>:<csignal>
<stdarg.h>:<cstdarg>
<stddef.h>:<cstddef>
<stdio.h>:<cstdio>
<stdlib.h>:<cstdlib>
<string.h>:<cstring>
<time.h>:<ctime>
<wchar.h>:<cwchar>
<wctype.h>:<cwctype>
<uchar.h>:<cuchar>
<inttypes.h>:<cinttypes>
<stdint.h>:<cstdint>

maflcko reviewed (commented):
lgtm

hebasto commented on ci/test/01_base_install.sh:
```diff
@@ -96,6 +96,34 @@ if [[ "${RUN_TIDY}" == "true" ]]; then
 EOF
+  # Prefer C++ headers over C counterparts.
+  # See: https://github.com/include-what-you-use/include-what-you-use/blob/clang_21/iwyu_include_picker.cc#L587-L629.
+  sed -i "s|\"<assert.h>\", kPublic|\"<assert.h>\", kPrivate|g" /include-what-you-use/iwyu_include_picker.cc
```
I’ve been thinking about this as well and decided to leave this decision for follow-ups, as maintaining consistency would require the same change in other patches.

hebasto replied:
> edit: The full patch would be:

[Taken](https://github.com/bitcoin/bitcoin/pull/33810#issuecomment-3667066005).

maflcko commented:
review ACK 73439910a02c2c07de94ad3a78ac3421ba7218d9 💣

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 73439910a02c2c07de94ad3a78ac3421ba7218d9 💣
PFE0MXlQDERuLW7wBsykme9L5AeWZS925eSdxV7Izq+OJxtpt5xAMKk8uh/6CsYBai+qxn1q2NB9oowhf4qPAw==
```

</details>

hebasto committed:
iwyu: Add patch to prefer angled brackets over quotes for includes

hebasto committed:
iwyu: Add patch to prefer C++ headers over C counterparts

hebasto committed:
cmake: Make `codegen` target dependent on `generate_build_info`

hebasto committed:
cmake: Fix target name

The executable target run by ctest is `mptest`.

This change removes unnecessary steps when building the `codegen`
target.

hebasto committed:
ci: Add IWYU job

The change in `src/crypto/hex_base.cpp` is because GCC 14 is not
affected by an IWYU bug.
See: https://github.com/include-what-you-use/include-what-you-use/issues/1763.

hebasto committed:
iwyu, clang-format: Sort includes

maflcko commented:
Only change is replacing the sed commands with a large patch.

review ACK 56750c4f87d089c6a3f093eb2bf2edd07170d4a8 🌄

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 56750c4f87d089c6a3f093eb2bf2edd07170d4a8 🌄
Xa/HJod6nR3+9bZFBU0zofE9PFyrRUFTjEk24LOKmdHtq/8KRknE4YKBRw4ac5u4m9HCwWeM+3iD8x2R8tfPCw==
```

</details>

sedited reviewed (approved):
ACK 56750c4f87d089c6a3f093eb2bf2edd07170d4a8
